Builds up RoadGeometry out of parsed information.

francocipollone opened the pull request:
# 🎉 New feature

Related to https://github.com/maliput/maliput_sparse/issues/34

## Summary

 - Brings Lane and Segment structs from maliput_osm given that they are not backend dependant
 - Creates a RoadGeometry loader that uses the parsed information(Filled up segment structs) and the builder api.
   - This implementation is brought from maliput_osm's road geometry builder as it is not backend dependant.
 - Propose an abstract Parser class that should be implemented by the backends, for parsing the correspondent underlying format and filling up the Lane and Segment structs.

This is a base proposal on https://github.com/maliput/maliput_sparse/issues/34, once this is given the thumbs up the code should be polished, tests should be changed/added and documentation should be added.

https://github.com/maliput/maliput_osm/pull/30 matches these changes here added.

## Checklist
- [x] Signed all commits for DCO
- [ ] Added tests
- [ ] Added example and/or tutorial
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] Consider updating Python bindings (if it affects the public API)

**Note to maintainers**: Remember to use **Squash-Merge** and edit the commit message to match the pull request summary while retaining `Signed-off-by` messages.

francocipollone commented:
As this is a proposal the code might change, however, the essence will remain, for so, I'd like a general review from you @agalbachicar on this.

The bottom line here is to move up to maliput_sparse all that is not backend dependent and that could easier the life to the backend's creators. In particular https://github.com/maliput/maliput_osm/pull/30 already shows how the implementation can be reduced to mainly just a parser for the underlying format and fill up some segment structs.

agalbachicar commented on include/maliput_sparse/parser/parser.h:
```diff
@@ -0,0 +1,63 @@
+  Parser() = default;
+  virtual ~Parser() = default;
+
+  /// Gets the segments parsed from the input description.
+  const std::unordered_map<Segment::Id, Segment>& GetSegments() const {return DoGetSegments();}
```
I think it is a bit premature to move things here. This is an example of that. We would likely need also the Junction information and we are not even configuring other RoadNetwork entities (e.g. Intersections, Phases, etc.). The Parser should evolve to to provide all of them most likely.

I would refrain adding this class for the moment.

agalbachicar commented on include/maliput_sparse/parser/lane.h:
```diff
@@ -0,0 +1,69 @@
+namespace parser {
+
+/// Contains the characteristics of a lane.
+struct Lane {
```
OSM itself has a width and a set of waypoints instead of a left and right sets. This struct is not generic at a glance.

We are also missing the most complicated cases in which surface information (e.g. CRG files) are provided. There is no support for that in these structs. This point is not a strong argument yet as we are not able to support it in any part of this implementation yet.

francocipollone replied:
> This struct is not generic at a glance.

I think that these structs should contain what we can handle with the builder api because at the end of the day the backend's creator must match the information from the parser to the builder. How can we make them even more generic?

agalbachicar replied:
Perhaps the linestring construction should happen outside this struct. The first time I read this PR, I made a mental model and intent to make this generic for other sparse formats, so restricting at the "parser" level the geometric model when the format is unknown it doesn't sound a good approach to me.

I may be missing context, then I think it's a matter of facing the use case in which another geometry model is required and adapting the code accordingly.

agalbachicar commented on include/maliput_sparse/parser/lane.h:
```diff
@@ -0,0 +1,69 @@
+#include "maliput_sparse/geometry/line_string.h"
+
+namespace maliput_sparse {
+namespace parser {
```
Why change it to parser from loader?

francocipollone replied:
It was under `maliput_osm/osm`.

agalbachicar reviewed (changes requested):
Some comments for discussion.

francocipollone commented:
Now that junctions are populated and the rest of the entities are also added to the road network I moved forwarded and continued this work.

This PR pairs with https://github.com/maliput/maliput_osm/pull/30

@agalbachicar  PTAL

francocipollone commented:
Once approved, I will fill up an issue for adding tests for the road network loader here proposed.

agalbachicar commented on include/maliput_sparse/loader/builder_configuration.h:
```diff
@@ -0,0 +1,74 @@
+
+/// Holds the configuration parameters for the builder.
+struct BuilderConfiguration {
+  /// Create a BuilderConfiguration from a string dictionary.
+  static BuilderConfiguration FromMap(const std::map<std::string, std::string>& config);
+
+  /// @details The keys of the map are listed at @ref builder_configuration_keys.
+  /// @returns A string-string map containing the builder configuration.
+  std::map<std::string, std::string> ToStringMap() const;
```
I think that we should make these free functions to keep the style of other repositories.
Consider deleting constructors if you want to enforce a specific way of contring from a map this struct.

agalbachicar replied:
It's a nit.
The comment was to keep PODs as structs or convert them to classes to have functions like these.

francocipollone replied:
Ah I see, thanks for the clarification. Let me move forward with this PR and create an issue for improving the style on that.

francocipollone replied:
https://github.com/maliput/maliput_sparse/issues/44

agalbachicar commented on include/maliput_sparse/loader/road_geometry_loader.h:
```diff
@@ -0,0 +1,58 @@
+namespace maliput_sparse {
+namespace loader {
+
+class RoadGeometryLoader {
```
Nit: missing class docstring.

francocipollone replied:
Done.

agalbachicar commented on include/maliput_sparse/parser/lane.h:
```diff
@@ -0,0 +1,80 @@
+  };
+  bool operator==(const LaneEnd& other) const;
+
+  // The lanelet id.
+  std::string lane_id;
+  // The endpoint of the lanelet.
+  Which end;
```
Shall we add default values?

francocipollone replied:
Done.

agalbachicar commented on include/maliput_sparse/parser/segment.h:
```diff
@@ -0,0 +1,57 @@
+  bool operator==(const Segment& other) const { return id == other.id && lanes == other.lanes; }
+
+  /// Id of the segment.
+  Id id;
```
Add `{}` please.

agalbachicar reviewed (changes requested):
A few comments to go.

agalbachicar reviewed (approved):
LGTM
